### DISCUSSION TITLE
Deleting the installed files and files from pervious version are not restored if the upgrade installation was failed.

### DISCUSSION THREAD

**Discussion by EiMaw** [wixtoolset/issues#6527] [Questions] [answered]
hi,
I have a bootstrapper app which bundle 3 msi (FirstPackage.msi, SecondPackage.msi and ThirdPackage.msi) and one custom action project (MyCustomAction).
Each msi also included the custom action in Product.wxs to call the method from MyCustomAction project.
Each msi will deploy the files during installation.

I have performed the upgrade installation with version 2.0.0.0 because I have already installed the first version 1.0.0.0.
During upgrade installation, ThirdPackage.msi was failed to install due to the error from the method of custom action and installation was failed.
FirstPackage and SecondPackage folder including files were deleted in the installed location when the upgrade installation was failed.
Old files from previous version (1.0.0.0) of FirstPackage and SecondPackage were not restored although I have used the Schedule="afterInstallExecute" in MajorUpgrade in each Product.wxs in all msi project. 

<MajorUpgrade DowngradeErrorMessage="A newer version of [ProductName] is already installed." AllowSameVersionUpgrades="yes"  Schedule="afterInstallExecute" />

![image](https://user-images.githubusercontent.com/87958309/126954024-e3ef8c6b-8699-4aea-8831-c29879ba418a.png)

![image](https://user-images.githubusercontent.com/87958309/126954546-ad76d8bf-3fc5-48a1-abce-445e24b7a04c.png)

Does the uninstallation (deleting all the installed files) perform if the upgrade installation was failed?

How can I achieve to restore the files or everything from old version automatically to go back to original state (before upgrade) if the upgrade installation was failed?

**Comment by robmen** [comment, marked as answer]
The fix is to repair the v1 bundle (there are other more magical mechanisms that would be desirable but all of them have a fatal flaw in the general case, so repair is the best option we are left with that works universally).

There is a feature request out there to automatically repair the old bundle in case of upgrade failure in a future version of the WiX Toolset. It is a non-trivial amount of work to do well in the general case.
